interoptest/src/testcoordinator/interoptestservice/interop_test_service.go

```diff
@@ -136,44 +139,100 @@ type ServiceImpl struct {
 	mu                 sync.Mutex
 	registeredServices map[string][]*interop.Service
 	sink               *receiver.TestCoordinatorSink
+	testSuites         map[*interop.Service][]*interop.ServiceHop
+	results            []*interop.TestResult
 }
 
 // NewService returns a new ServiceImpl with the given registered services.
 func NewService(services map[string][]*interop.Service, sink *receiver.TestCoordinatorSink) *ServiceImpl {
-	return &ServiceImpl{registeredServices: services, sink: sink}
+	return &ServiceImpl{registeredServices: services, sink: sink, testSuites: loadTestSuites()}
 }
 
+// SetRegisteredServices sets the registered services to the given one.
 func (s *ServiceImpl) SetRegisteredServices(services map[string][]*interop.Service) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	s.registeredServices = services
 }
 
+// Result returns all the test results.
 func (s *ServiceImpl) Result(ctx context.Context, req *interop.InteropResultRequest) (*interop.InteropResultResponse, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	reps := &interop.InteropResultResponse{
 		Id:     req.Id,
 		Status: &interop.CommonResponseStatus{Status: interop.Status_SUCCESS},
-		// TODO: store and return cached result
-		Result: []*interop.TestResult{},
+		Result: s.results,
 	}
 	return reps, nil
 }
 
-// Runs the test asynchronously.
+// Run runs the test asynchronously.
 func (s *ServiceImpl) Run(ctx context.Context, req *interop.InteropRunRequest) (*interop.InteropRunResponse, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	id := rand.Int63()
-	// TODO: run all tests by sending out test execution requests
-	verifySpans(s.sink.SpansPerNode)
+	for svc, hops := range s.testSuites {
+		sender, _ := testexecutionservice.NewUnstartedSender(true, id, svc.Name, fmt.Sprintf("%s:%d", svc.Host, svc.Port), hops)
+		reps, err := sender.Start()
+
+		if err != nil {
+			s.results = append(s.results, getFailedResult(id, svc.Name, hops, nil))
+			continue
+		}
+
+		for _, status := range reps.Status {
+			if status.Status == interop.Status_FAILURE {
+				s.results = append(s.results, getFailedResult(id, svc.Name, hops, reps.Status))
+				continue
+			}
+		}
+
+		// TODO: consider using a semaphore instead of waiting, to make exporting deterministic
+		time.Sleep(90 * time.Second) // wait until all micro-services exported their spans
```

Comment:
it seems each testSuite is one testcase and hence 90 second sleep is after each testcase. By testcase I mean one request to Svc A containing hops (B, C, A) or any combination of hops.

Reply:
that will take a long time to run the test. we probably will have 20-30 testcases and it will ony increase over time.

Reply:
I reduced the waiting time to 20s. Unfortunately we cannot start verifying the traces until all spans are exported, otherwise we won't be able to reconstruct traces and will mistakenly mark the test case as "failed".

Reply:
as per our conversation you will create an issue to optimize it later.

Reply:
Chatted offline. Once all micro services added attributes to the spans, we can do the verification after all test cases finished. Added a TODO here and updated https://github.com/census-ecosystem/opencensus-experiments/issues/133.